modules/tensor_mechanics/src/materials/CZM/CZM3DCLaw.C

```diff
@@ -0,0 +1,169 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CZM3DCLaw.h"
+
+registerMooseObject("TensorMechanicsApp", CZM3DCLaw);
+
+template <>
+InputParameters
+validParams<CZM3DCLaw>()
+{
+  InputParameters params = validParams<CZMMaterialBase>();
+  params.addClassDescription("3DC cohseive law model, no damage");
+  params.addRequiredParam<Real>(
+      "normal_gap_at_maximum_normal_traction",
+      "the value of normal gap at which maximum normal traction is achieved");
+  params.addRequiredParam<Real>(
+      "tangential_gap_at_maximum_shear_traction",
+      "the value of tangential gap at which maximum shear traction is achieved");
+  params.addRequiredParam<Real>("maximum_normal_traction",
+                                "The maximum normal traction the interface can sustain");
+  params.addRequiredParam<Real>("maximum_shear_traction",
+                                "The maximum shear traction the interface can sustain");
+  params.addClassDescription("Simple Exponential cohseive law model, with damage");
+  return params;
+}
+
+CZM3DCLaw::CZM3DCLaw(const InputParameters & parameters)
+  : CZMMaterialBase(parameters), _deltaU0(3, 0), _maxAllowableTraction(3, 0)
+{
+
+  const_cast<std::vector<Real> &>(_deltaU0)[0] =
+      getParam<Real>("normal_gap_at_maximum_normal_traction");
+  const_cast<std::vector<Real> &>(_maxAllowableTraction)[0] =
+      getParam<Real>("maximum_normal_traction");
+
+  for (unsigned int i = 1; i < 3; i++)
+  {
+    const_cast<std::vector<Real> &>(_deltaU0)[i] =
+        getParam<Real>("tangential_gap_at_maximum_shear_traction") * std::sqrt(2);
+    const_cast<std::vector<Real> &>(_maxAllowableTraction)[i] =
+        getParam<Real>("maximum_shear_traction");
+  }
+}
+
+RealVectorValue
+CZM3DCLaw::computeTraction()
+{
+  RealVectorValue traction_local;
```

Comment:
Follow the variable naming convention standards for MOOSE:
https://mooseframework.org/framework_development/code_standards.html

Reply:
this is in fact a local variable for this function. The idea was that this function shall return only the traction vector in local coordinates and shall not member properties.  However, for stateful traction separation laws we need to modify member variables as they are material properties. You can look at my answer to your comments.